src/core/AudioEngine.cpp

```diff
@@ -35,37 +69,96 @@ namespace H2Core
 {
 
 
-AudioEngine* AudioEngine::__instance = nullptr;
-const char* AudioEngine::__class_name = "AudioEngine";
+inline int randomValue( int max )
+{
+	return rand() % max;
+}
+
+inline float getGaussian( float z )
+{
+	// gaussian distribution -- dimss
+	float x1, x2, w;
+	do {
+		x1 = 2.0 * ( ( ( float ) rand() ) / RAND_MAX ) - 1.0;
+		x2 = 2.0 * ( ( ( float ) rand() ) / RAND_MAX ) - 1.0;
+		w = x1 * x1 + x2 * x2;
+	} while ( w >= 1.0 );
+
+	w = sqrtf( ( -2.0 * logf( w ) ) / w );
+	return x1 * w * z + 0.0; // tunable
+}
 
 
-void AudioEngine::create_instance()
+/** Gets the current time.
+ * \return Current time obtained by gettimeofday()*/
+inline timeval currentTime2()
 {
-	if( __instance == nullptr ) {
-		__instance = new AudioEngine;
-	}
+	struct timeval now;
+	gettimeofday( &now, nullptr );
+	return now;
 }
 
+const char* AudioEngine::__class_name = "AudioEngine";
+
 AudioEngine::AudioEngine()
 		: Object( __class_name )
-		, __sampler( nullptr )
-		, __synth( nullptr )
+		, m_pSampler( nullptr )
+		, m_pSynth( nullptr )
 		, m_fElapsedTime( 0 )
+		, m_pMainBuffer_L( nullptr )
+		, m_pMainBuffer_R( nullptr )
+		, m_pAudioDriver( nullptr )
+		, m_pMidiDriver( nullptr )
+		, m_pMidiDriverOut( nullptr )
+		, m_State( STATE_INITIALIZED )
+		, m_pMetronomeInstrument( nullptr )
+		, m_nBufferSize( 0 )
+		, m_nPatternStartTick( -1 )
+		, m_nPatternTickPosition( 0 )
+		, m_nSongSizeInTicks( 0 )
+		, m_nRealtimeFrames( 0 )
+		, m_nAddRealtimeNoteTickPosition( 0 )
+		, m_nSongPos( -1 )
+		, m_nSelectedPatternNumber( 0 )
 {
-	__instance = this;
 	INFOLOG( "INIT" );
 
-	__sampler = new Sampler;
-	__synth = new Synth;
+	m_pSampler = new Sampler;
+	m_pSynth = new Synth;
+	
+	m_pEventQueue = EventQueue::get_instance();
+	
+	m_fMasterPeak_L = 0.0f;		///< Master peak (left channel)
+	m_fMasterPeak_R = 0.0f;		///< Master peak (right channel)
+	m_fProcessTime = 0.0f;		///< time used in process function
+	m_fMaxProcessTime = 0.0f;	///< max ms usable in process with no xrun
+	
+	srand( time( nullptr ) );
+
+	// Create metronome instrument
+	// Get the path to the file of the metronome sound.
+	QString sMetronomeFilename = Filesystem::click_file_path();
+	m_pMetronomeInstrument = new Instrument( METRONOME_INSTR_ID, "metronome" );
+	
+	InstrumentLayer* pLayer =  new InstrumentLayer( Sample::load( sMetronomeFilename ) );
+	InstrumentComponent* pCompo = new InstrumentComponent( 0 );
+	pCompo->set_layer(pLayer, 0);
+	m_pMetronomeInstrument->get_components()->push_back( pCompo );
+	m_pMetronomeInstrument->set_is_metronome_instrument(true);
+	
+	m_pPlayingPatterns = new PatternList();
+	m_pPlayingPatterns->setNeedsLock( true );
+	m_pNextPatterns = new PatternList();
+	m_pNextPatterns->setNeedsLock( true );
+	
+	m_AudioProcessCallback = &audioEngine_process;
 
 #ifdef H2CORE_HAVE_LADSPA
 	Effects::create_instance();
 #endif
 
 }
 
-
-
 AudioEngine::~AudioEngine()
 {
 	INFOLOG( "DESTROY" );
```

Comment:
How about checking for `! m_songNoteQueue.empty()`, `m_pPlayingPatterns != nullptr`, `m_pNextPatterns != nullptr`, and `m_pMetronomeInstrument != nullptr` in here and calling `destroy()` in case one of these conditions is false. This way we can ensure everything is clean after `AudioEngine` is destroyed regardless of the changes that might be applied to `Hydrogen::~Hydrogen` in the future.